temporalio/worker.py

```diff
@@ -0,0 +1,1107 @@
+                            thread_event=threading.Event(), async_event=None
+                        ),
+                        sync=False,
+                        last_heartbeat_task=None,
+                    )
+                    activity.task = asyncio.create_task(
+                        self._run_activity(task.task_token, task.start, activity)
+                    )
+                    self._running_activities[task.task_token] = activity
+                elif task.HasField("cancel"):
+                    self._cancel_activity(task.task_token, task.cancel)
+                else:
+                    logger.warning("unrecognized activity task: %s", task)
+            except temporalio.bridge.worker.PollShutdownError:
+                return
+            except Exception as err:
+                # Should never happen
+                logger.exception(f"Activity runner failed: {err}", extra={"error": err})
+
+    async def _run_workflows(self) -> None:
+        raise NotImplementedError
+
+    def _cancel_activity(
+        self, task_token: bytes, cancel: temporalio.bridge.proto.activity_task.Cancel
+    ) -> None:
+        activity = self._running_activities.get(task_token)
+        if not activity:
+            logger.warning("Cannot find activity to cancel for token %s", task_token)
+            return
+        logger.debug("Cancelling activity %s, reason: %s", task_token, cancel.reason)
+        activity.cancel(cancelled_by_request=True)
+
+    def _heartbeat_activity(self, task_token: bytes, *details: Any) -> None:
+        # We intentionally make heartbeating non-async, but since the data
+        # converter is async, we have to schedule it
+        logger = temporalio.activity.logger
+        activity = self._running_activities.get(task_token)
+        if activity:
+            activity.current_heartbeat_seq += 1
+            heartbeat_seq = activity.current_heartbeat_seq
+            activity.last_heartbeat_task = asyncio.create_task(
+                self._heartbeat_activity_async(
+                    logger, task_token, heartbeat_seq, *details
+                )
+            )
+
+    async def _heartbeat_activity_async(
+        self,
+        logger: logging.Logger,
+        task_token: bytes,
+        heartbeat_seq: int,
+        *details: Any,
+    ) -> None:
+        activity = self._running_activities.get(task_token)
+        # Bail if not the latest
+        if not activity or activity.current_heartbeat_seq != heartbeat_seq:
+            return
+        try:
+            heartbeat = temporalio.bridge.proto.ActivityHeartbeat(task_token=task_token)
+            if details:
+                heartbeat.details.extend(
+                    temporalio.bridge.worker.payloads_to_core(
+                        await self._config["client"].data_converter.encode(details)
+                    )
+                )
+            # Bail if not the latest
+            if activity.current_heartbeat_seq != heartbeat_seq:
```

Comment:
I'd treat this `try` block as a single operation, encode and send to core, no bailing in the middle.

Reply:
You can use a single consumer (coroutine) per activity to avoid this, it'll make heartbeating much simpler.

Reply:
I will change to treat this as a single uninterruptible task. I didn't want to start a heartbeater coroutine compared to waiting until ever needed (though I know coroutines are cheap) because I think it adds additional complexities on completion, but I'll investigate.

Reply:
Changed.